File: dynamic_programming/linear_dp/single_chain/0005_longest_palindromic_substring.js
/*
Given a string s, return the longest palindromic substring in s.
*/

// https://leetcode.com/problems/longest-palindromic-substring/description/

/* dynamic programming
let dp[i][j] be true if s[i:j] is a palindrome
1) if s[i] === s[j]:
  1. if j - i < 3, dp[i][j] = true
  2. if j - i >= 3, dp[i][j] = dp[i + 1][j - 1]
2) if s[i] !== s[j], dp[i][j] = false
*/
var longestPalindrome = function (s) {
  if (s.length < 2) return s;
  let dp = new Array(s.length)
    .fill(0)
    .map(() => new Array(s.length).fill(false));
  let res = "";
  for (let i = s.length - 1; i >= 0; i--) {
    for (let j = i; j < s.length; j++) {
      dp[i][j] = s[i] === s[j] && (j - i < 3 || dp[i + 1][j - 1]);
      if (dp[i][j] && j - i + 1 > res.length) {
        res = s.slice(i, j + 1);
      }
    }
  }
  return res;
};

// center spread
var longestPalindrome = function (s) {
  if (s.legnth <= 2) return s;
  let res = "";
  for (let i = 0; i < s.length; i++) {
    centerSpread(i, i);
    centerSpread(i, i + 1);
  }
  function centerSpread(start, end) {
    while (start >= 0 && end < s.length && s[start] == s[end]) {
      start--;
      end++;
    }
    if (end - start - 1 > res.length) {
      res = s.slice(start + 1, end);
    }
  }
  return res;
};

File: dynamic_programming/linear_dp/single_chain/0005_longest_palindromic_substring.py
# Given a string s, return the longest palindromic substring in s.

# https://leetcode.com/problems/longest-palindromic-substring/
# similar
# LC 647 https://leetcode.com/problems/palindromic-substrings/

# center spread
class Solution:
    def longestPalindrome(self, s: str) -> str:
        l = len(s)
        if l <= 1: return s
        res = ''
        for i in range(l):
            temp1 = self.centerSpread(s, i, i)
            if len(temp1) > len(res):
                res = temp1
            temp2 = self.centerSpread(s, i, i + 1)
            if len(temp2) > len(res):
                res = temp2
        return res
    # find longest palingromic substring with certain start and end
    def centerSpread(self, s: str, start: int, end: int) -> str:
        l = len(s)
        while start >= 0 and end < l and s[start] == s[end]:
            start -= 1
            end += 1
        return s[start + 1 : end]
        
# dynamic programming
# let dp[i][j] be true if s[i:j] is a palindrome
# 1) if s[i] == s[j]:
#   if j - i <= 2, dp[i][j] is True
#   if j - i > 2, dp[i][j] = dp[i + 1][j - 1]
# 2) if s[i] != s[j]: dp[i][j] is False
# 
class Solution:
    def longestPalindrome(self, s: str) -> str:
        l = len(s)
        if l <= 1: return s
        dp = [[False] * l for _ in range(l)]
        res = ''

        for i in range(l - 1, -1, -1):
            for j in range(i, l):
                dp[i][j] = s[i] == s[j] and (j - i <= 2 or dp[i + 1][j - 1])
                if dp[i][j] and j - i + 1 > len(res):
                    res = s[i : j + 1]
        return res